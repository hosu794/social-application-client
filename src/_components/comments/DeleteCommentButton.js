import React from "react";

import { commentActions } from "../../_actions";

import { useDispatch } from "react-redux";

function DeleteCommentButton({ handleDeleteComment }) {
  return (
    <button onClick={handleDeleteComment} className="button is-danger">
      Delete
    </button>
  );
}

export default DeleteCommentButton;
