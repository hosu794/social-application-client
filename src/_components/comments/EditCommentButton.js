import React from "react";

import { Link } from "react-router-dom";

function EditCommentButton({ id }) {
  let commentEditUrl = `/comment/edit/${id}`;

  return (
    <Link to={commentEditUrl}>
      <button className="button is-info">Edit</button>
    </Link>
  );
}

export default EditCommentButton;
