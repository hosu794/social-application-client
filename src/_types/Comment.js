import PropTypes from "prop-types";
import { Component } from "react";
import { IUser } from "./";

export const IComment = (Component.PropTypes = {
  id: PropTypes.number.isRequired,
  body: PropTypes.string.isRequired,
  createdBy: PropTypes.instanceOf(IUser),
});
