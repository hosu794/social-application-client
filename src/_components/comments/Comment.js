import React from "react";

import { useDispatch, useSelector } from "react-redux";
import DeleteCommentButton from "./DeleteCommentButton";
import EditCommentButton from "./EditCommentButton";
import { commentActions } from "../../_actions";
import PropTypes from "prop-types";
import { IComment, IUser } from "../../_types";

function Comment({ comment }) {
  const dispatch = useDispatch();

  const user = useSelector((state) => state.user.user);
  const isUserExist = user ? user.id : null;
  const isUserIdentificationEqualStoryCreator =
    isUserExist === comment.createdBy.id;

  function handleDeleteComment() {
    dispatch(commentActions.deleteComment(comment.id));
  }

  return (
    <div className="column">
      <div className="card">
        <nav className="level">
          <div class="level-item has-text-centered">
            <div className="is-size-5">{comment.body}</div>
          </div>
        </nav>
        <nav className="level">
          <div class="level-item has-text-centered">
            <div className="is-size-6">
              created by <b>{comment.createdBy.username}</b>
            </div>
          </div>
          <div className="level-item has-text-centered">
            {isUserIdentificationEqualStoryCreator ? (
              <DeleteCommentButton handleDeleteComment={handleDeleteComment} />
            ) : null}
          </div>
          <div className="level-item has-text-centered">
            {isUserIdentificationEqualStoryCreator ? (
              <EditCommentButton id={comment.id} />
            ) : null}
          </div>
        </nav>
      </div>
    </div>
  );
}

Comment.propTypes = {
  comment: PropTypes.instanceOf(IComment).isRequired,
  user: PropTypes.instanceOf(IUser),
  isUserExist: PropTypes.number,
  isUserIdentificationEqualStoryCreator: PropTypes.bool,
};

Comment.defaultProps = {
  comment: null,
  user: {},
  isUserExist: undefined,
  isUserIdentificationEqualStoryCreator: false,
};

export default Comment;
