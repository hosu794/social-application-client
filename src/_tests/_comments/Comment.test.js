import React from "react";

import configureStore from "redux-mock-store";
import { useDispatch } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Comment from "../../_components/comments/Comment";
import { mount } from "enzyme";

import { Provider } from "react-redux";

import EditCommentButton from "../../_components/comments/EditCommentButton";
import DeleteCommentButton from "../../_components/comments/DeleteCommentButton";

const mockStore = configureStore([]);

let store = mockStore({
  comment: {
    content: [
      {
        id: 1,
        body: "Some body",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    ],
  },
  user: {
    isUverLovedStory: true,
    user: {
      id: 12,
      name: "Joe Doe",
      username: "exampleusername",
      downloadAvatar: "dsdasdasdasdasdadas.png",
    },
  },
  authentication: {
    loggenIn: true,
  },
  stories: {
    content: [],
    currentStory: {
      id: 1,
      title: "Story Title",
      description: "Story description",
      totalLoves: 12,
      body: "<p>Simple Story</p>",
      createdBy: {
        id: 12,
        name: "Joe Doe",
        username: "exampleusername",
        downloadAvatar: "dsdasdasdasdasdadas.png",
      },
      topic: {
        id: 12,
        title: "Topic Example Title",
        description: "Description Example description",
        createdAt: "Dummy date",
        createdBy: {
          id: 12,
          name: "Joe Doe",
          username: "exampleusername",
          downloadAvatar: "dsdasdasdasdasdadas.png",
        },
      },
    },
  },
});

let mockComment = {
  id: 1,
  body: "Some body",
  createdBy: {
    id: 12,
    name: "Joe Doe",
    username: "exampleusername",
    downloadAvatar: "dsdasdasdasdasdadas.png",
  },
};

describe("Test for the Comment Component", () => {
  let TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Comment comment={mockComment} />
      </Provider>
    </MemoryRouter>
  );

  let wrapper = mount(<TestingComponent />);

  test("should render component correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(Comment).toBeDefined();
  });

  test("should render correctly component props", () => {
    let element;

    element = wrapper.find(".is-size-5").first().text();
    expect(element).toEqual(mockComment.body);

    element = wrapper.find(".is-size-6").first().text();

    let CREATED_BY = "created by ";

    expect(element).toEqual(CREATED_BY + mockComment.createdBy.username);
  });

  test("should component containst EditCommentButton", () => {
    let element = wrapper.exists(EditCommentButton);
    expect(element).toEqual(true);
  });

  test("should component constainst DeleteCommentButton", () => {
    let element = wrapper.exists(DeleteCommentButton);

    expect(element).toEqual(true);
  });
});
