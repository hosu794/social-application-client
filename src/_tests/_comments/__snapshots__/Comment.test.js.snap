// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Comment Component should render component correctly 1`] = `ReactWrapper {}`;
